lib/googleauth/external_account.rb

```diff
@@ -0,0 +1,62 @@
+# Copyright 2015 Google, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "time"
+require "googleauth/credentials_loader"
+require "googleauth/external_account/aws_credentials"
+
+module Google
+  # Module Auth provides classes that provide Google-specific authorization
+  # used to access Google APIs.
+  module Auth
+    # Authenticates requests using External Account credentials, such
+    # as those provided by the AWS provider.
+    class ExternalAccountCredentials
+      attr_reader :project_id
+      attr_reader :quota_project_id
+
+      # Create a ExternalAccountCredentials
+      #
+      # @param json_key_io [IO] an IO from which the JSON key can be read
+      # @param scope [string|array|nil] the scope(s) to access
+      def self.make_creds options = {}
+        json_key_io, scope = options.values_at :json_key_io, :scope
+
+        raise "a json file is required for external account credentials" unless json_key_io
+        user_creds = read_json_key json_key_io
+
+        Google::Auth::ExternalAccount::AwsCredentials.new(
```

Comment:
There are a few different types of external account credential files that exist right now.  Let's make sure the file is intended to be AWS credentials before we use that class to create them, and raise an "unknown credential" error until we've implemented them.  Use either of these methods:

In GoLang, we make sure that the environment ID of the credential source is "AWS1" before we create AWS credentials (and could be AWS2, AWS3, etc for future versions of these credentials)

In Python, we make sure that the subject token type of the credential source is "urn:ietf:params:aws:token-type:aws4_request"

Reply:
Fixed in 994456a427d3f12e4b29bf5f5397cdb49605fe2a